app/dao/unsubscribe_request_dao.py

```diff
@@ -80,3 +105,20 @@ def create_unsubscribe_request_reports_dao(unsubscribe_request_report):
 def update_unsubscribe_request_report_processed_by_date_dao(report, report_has_been_processed):
     report.processed_by_service_at = datetime.utcnow() if report_has_been_processed else None
     db.session.add(report)
+
+
+@autocommit
+def assign_unbatched_unsubscribe_requests_to_report_dao(report_id, service_id, earliest_timestamp, latest_timestamp):
+    """
+    This method retrieves all the un-batched unsubscribe requests received before or on
+    the report's latest_timestamp and sets their unsubscribe_request_report_id to the report_i
+    """
+    unbatched_unsubscribe_requests = UnsubscribeRequest.query.filter(
+        UnsubscribeRequest.unsubscribe_request_report_id.is_(None),
+        UnsubscribeRequest.service_id == service_id,
+        UnsubscribeRequest.created_at >= earliest_timestamp,
+        UnsubscribeRequest.created_at <= latest_timestamp,
+    ).all()
+    for unsubscribe_request in unbatched_unsubscribe_requests:
+        unsubscribe_request.unsubscribe_request_report_id = report_id
+        db.session.add(unsubscribe_request)
```

Comment:
I’m wondering if it’s better to batch up all the unbatched requests, rather than leave any that arrive after `latest_timestamp` dangling

Pro: you will get a download of everything you need to action
Con: the `count` might not quite match up to what you see on the page

Reply:
I wondered about that as well, the `latest_timestamp` is also the `earliest_timestamp` of the prospective next report so the un-batched reports that arrive soon after that will go into that next report.
One other con is that we may have an inconsistency where reports either have requests in them that are beyond their `latest_timestamps` or not have requests that should have been in them considering their `earliest_timestamps`.
My thinking on the issue is that this is one we can revisit once we get an idea of how users get on with the feature. It would be a simple fix whichever way we go.

Reply:
Yeah I’ve had a deeper think about this and I think the way you have it is right. Admin `POST`s `earliest_timestamp` and `latest_timestamp` and the API returns what’s between those two and nothing more.